feat(filter): support phrase matching in filter_by

felipehertzer opened the pull request:
## Change Summary

Hi, I'm having an issue filtering documents when using a compound word such as **"Formula One"**.

Currently, filtering behaves as follows:

* **Query:** `filter_by: text: Formula One`
* **Result:** Returns documents that contain both words separately, for example:

> "The scientist created a new formula, and this was just one of many groundbreaking discoveries in the lab."

What I expected was to only retrieve documents that contain the exact phrase:

> "Lewis Hamilton has won multiple Formula One World Championships."

I noticed that Typesense already supports the use of backticks to escape characters, so as an initial attempt, I reused the same syntax:

```text
filter_by: 'text: `Formula One`'
```

Close:
#2205 #2198 #1219

Thanks.

## PR Checklist
<!--- Put an `x` inside the box : -->
- [x] I have read and signed the [Contributor License Agreement](https://forms.gle/PZyiY5N2GDQU8GsV9).

kishorenc commented:
Thank you, this will be a useful feature. We will be reviewing the PR next week.

felipehertzer commented:
Hey @kishorenc,

I’ve fixed the test errors and added additional test cases as well. There is still one failing test on my machine, which I believe may be related to a pointer or an environment-specific issue in `CollectionSortingTest.TestVectorQueryDistanceThresholdSorting`.

```
Expected equality of these values:
  0.07853113859891891
    Which is: 0.0785311
  res["hits"][0]["vector_distance"].get<float>()
    Which is: 0.0785313
```

happy-san commented:
Also, we have to support lazy evaluation of the filter_by expression. You can refer [this test](https://github.com/typesense/typesense/blob/v30/test/filter_test.cpp#L486-L502) for a better understanding. 
The data in the test won't induce lazy evaluation behavior. Maybe we should have documents having following values:
```
foo
foo bar baz
foo bar
bar
foo bar baz
baz
foo baz bar
foo bar baz
```
and the query:
```
q: foo
filter_by: field: "foo bar baz"
enable_lazy_filter: true
```
which should match 3 documents. Let me know if something doesn't make sense.
